apis/duck/v1beta1/source_types.go

```diff
@@ -0,0 +1,174 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"context"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	apisv1alpha1 "knative.dev/pkg/apis/v1alpha1"
+)
+
+// Source is an Implementable "duck type".
+var _ duck.Implementable = (*Source)(nil)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Source is the minimum resource shape to adhere to the Source Specification.
+// This duck type is intended to allow implementors of Sources and
+// Importers to verify their own resources meet the expectations.
+// This is not a real resource.
+// NOTE: The Source Specification is in progress and the shape and names could
+// be modified until it has been accepted.
+type Source struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SourceSpec   `json:"spec"`
+	Status SourceStatus `json:"status"`
+}
+
+type SourceSpec struct {
+	// Sink is a reference to an object that will resolve to a domain name or a
+	// URI directly to use as the sink.
+	Sink apisv1alpha1.Destination `json:"sink,omitempty"`
+
+	// CloudEventOverrides defines overrides to control the output format and
+	// modifications of the event sent to the sink.
+	// +optional
+	CloudEventOverrides CloudEventOverrides `json:"ceOverrides,omitempty"`
+}
+
+type CloudEventOverrides struct {
+	// OutputFormat describes the CloudEvent output format the source
+	// should send events in. All formats are over HTTP.
+	// Defaults to binary.
+	// +optional
+	OutputFormat OutputFormatType `json:"outputFormat,omitempty"`
+
+	// Extensions will specify what is added to an event as an extension.
+	// Each key-value pair is added independently.
+	// +optional
+	Extensions map[string]string `json:"extensions,omitempty"`
+}
+
+// SourceStatus shows how we expect folks to embed Addressable in
+// their Status field.
+type SourceStatus struct {
+	// inherits duck/v1beta1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last
+	//   processed by the controller.
+	// * Conditions - the latest available observations of a resource's current
+	//   state.
+	Status `json:",inline"`
+
+	// SinkURI is the current active sink URI that has been configured for the
+	// Source.
+	// +optional
+	SinkURI *apis.URL `json:"sinkUri,omitempty"`
+}
+
+var (
+	// Verify Source resources meet duck contracts.
+	_ duck.Populatable = (*Source)(nil)
+	_ apis.Listable    = (*Source)(nil)
+)
+
+const (
+	// SourceConditionSinkProvided has status True when the Source
+	// has been configured with a sink target that is resolvable.
+	SourceConditionSinkProvided apis.ConditionType = "SinkProvided"
```

Comment:
Could / should there be a SourceConditionReady that all should use as the top level Happy condition?

Reply:
Added an `IsReady` method. I can't assume Ready because we will support Job based sources.

Reply:
Could add a `Succeeded` method